### ISSUE TITLE
System memory leak when using different input size of torch.nn.Conv3d

### ISSUE THREAD

**Issue by gerbenvv** [pytorch/pytorch#104701]
### 🐛 Describe the bug

There is a system memory leak when using different input sizes to `torch.nn.Conv3d` on the GPU. A very simple script to reproduce the issue is:

```python
import gc

import psutil
import torch
import torch._C
import torch._dynamo
import torch.backends.cudnn as cudnn


class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()

        self.convolution = torch.nn.Sequential(
            torch.nn.Conv3d(
                in_channels=1,
                out_channels=32,
                kernel_size=1,
                dilation=1,
                bias=False,
            ),
            torch.nn.Conv3d(
                in_channels=32,
                out_channels=1,
                kernel_size=1,
                dilation=1,
                bias=False,
            ),
        )

    def forward(self, input, labels):
        output = self.convolution(input)

        loss = torch.abs(output - labels).mean()

        return loss


class Trainer:
    def train(self) -> None:
        # Set our device.
        device = torch.device("cuda:0")

        # Set some training flags.
        cudnn.benchmark = False
        cudnn.deterministic = False

        # Create module.
        module = Model()
        module.to(device=device)

        # Put model in train mode.
        module.train()

        # Create optimizer.
        optimizer = torch.optim.Adam(module.parameters(), lr=1e-5)
        optimizer.zero_grad()

        # Train for a number of steps.
        for step in range(1, 1_000_000):
            if step % 10 == 0:
                self._report_leak_stats(step)

            input, labels = self._get_example(device)

            # Run model.
            loss = module(input, labels)

            # Raise on nan.
            if torch.isnan(loss).item():
                raise RuntimeError("Encountered nan loss during training.")

            # Do backwards pass.
            loss.backward()

            # Step optimizer and zero gradients after.
            optimizer.step()
            optimizer.zero_grad()

    def _get_example(self, device):
        depth = int(torch.randint(10, 80, size=()))
        height = int(torch.randint(10, 80, size=()))
        width = int(torch.randint(10, 80, size=()))

        return (
            torch.rand((1, 1, depth, height, width), dtype=torch.float32, device=device),
            torch.rand((1, 1, depth, height, width), dtype=torch.float32, device=device),
        )

    def _report_leak_stats(self, step):
        gc.collect()
        torch._C._cuda_clearCublasWorkspaces()
        torch._dynamo.reset()
        gc.collect()
        torch.cuda.empty_cache()
        gc.collect()

        print(
            "step",
            step,
            "residual memory",
            psutil.Process().memory_info().rss / 1e6,
            "MB",
        )


def main():
    Trainer().train()


if __name__ == "__main__":
    main()
```

When running it will show the residual memory of the process which keeps increasing over time. This is not the case for fixed inputs. The memory leak does not appear to be Python objects (I checked using various profilers). The memory leak seems to be affected by the filter size (and corresponding padding). Bigger filter sizes create bigger leaks.

The leak is fairly small per step (about 4 MB is leaked per 10 steps) but is really problematic when using DDP with multiple processes as it multiplies. Also the leak seems to be bigger for more convolutions.

Took me a very long time to track this leak down to make a minimal example :sweat_smile: . It restricts us from training a 3D model to convergence that takes variable sized inputs (3D scans). I feel like there could be other use cases where the input is variable.

### Versions

```                                                                                                                                                                                                                                  
PyTorch version: 2.0.1+cu117                                                                                                                                                                                                                                                   
Is debug build: False                                                                                                                                                                                                                                                          
CUDA used to build PyTorch: 11.7                                                                                                                                                                                                                                               
ROCM used to build PyTorch: N/A                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                               
OS: Ubuntu 22.04.1 LTS (x86_64)                                                                                                                                                                                                                                                
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0                                                                                                                                                                                                                           
Clang version: 14.0.0-1ubuntu1                                                                                                                                                                                                                                                 
CMake version: version 3.26.0                                                                                                                                                                                                                                                  
Libc version: glibc-2.35                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                               
Python version: 3.10.9 (main, Mar 29 2023, 10:13:28) [GCC 11.3.0] (64-bit runtime)                                                                                                                                                                                             
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35                                                                                                                                                                                                                 
Is CUDA available: True                                                                                                                                                                                                                                                        
CUDA runtime version: Could not collect                                                                                                                                                                                                                                        
CUDA_MODULE_LOADING set to: LAZY                                                                                                                                                                                                                                               
GPU models and configuration:                                                                                                                                                                                                                                                  
GPU 0: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 1: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 2: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 3: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 4: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 5: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 6: NVIDIA GeForce RTX 3090                                                                                                                                                                                                                                                 
GPU 7: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.60.11
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          64
On-line CPU(s) list:             0-63
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7302 16-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       2
Stepping:                        0
Frequency boost:                 enabled
CPU max MHz:                     3000.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        5999.84
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                  AMD-V
L1d cache:                       1 MiB (32 instances)
L1i cache:                       1 MiB (32 instances)
L2 cache:                        16 MiB (32 instances)
L3 cache:                        256 MiB (16 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-15,32-47
NUMA node1 CPU(s):               16-31,48-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy==1.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] torch==2.0.1
[pip3] torch-summary==1.4.5
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] Could not collect
```

**Comment by gerbenvv** [comment]
This could be a duplicate of https://github.com/pytorch/pytorch/issues/98688

Anyway, gives you another reproduction script/case of it.

**Comment by eqy** [comment]
#104369 has been opened to work around the memory usage issue.

**Comment by Draqun** [comment]
Hello guys,

I made some independent confirmation on current PyTorch. Also I prepared a controlled experiment and source-level analysis which shows where the host memory goes and why it is basically unbounded for long-running multi-threaded inference server. (We hit this problem on production. We mitigated it on our side, but I think it is worth to fix it upstream).

**TL;DR**
- Host RSS grows with number of distinct conv input shapes (not with #calls or size of buffer). On CUDA it is cuDNN v8 plan cache; `torch.backends.cudnn.enabled=False` removes this shape-scaling completely.
- PyTorch *does* bound this cache with LRU (`TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT`, default 10000 entries ≈ 2 GiB), but two things make it effectively unbounded for server:
  1. cache is thread-local, so budget is per thread — total host memory usage scales with number of threads running convolutions. There is no process-wide limit;
  2. since cuDNN 9.10+, thread-local caches are intentionally leaked (never freed) to prevent teardown segfault — so when threads exit, they never reclaim their cache.

### Environment

```
torch 2.10.0+cu128   CUDA 12.8   cuDNN 9.10.2 (91002)
GPU: NVIDIA RTX A500 (driver 580.159.03)   Linux
```

### Where is it in the code

`aten/src/ATen/native/cudnn/Conv_v8.cpp`:

- `getLRUCacheLimit()` — default `10000` ("roughly corresponds to 2GiB assuming 200KiB per ExecutionPlan"); `0` = no limit, negative = no caching; env `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT`.
- `struct BenchmarkCache` — `unordered_map<key, (ExecutionPlan, list-iterator)>` + `engine_cache_order` list = an LRU; `update()` evicts the back only once `size() >= limit`.
- `_get_benchmark_cache()` — the cache is `static thread_local … = new BenchmarkCache(...)`, with this comment:

  ```cpp
  // @eqy: use thread local caches as cuDNN Execution Plans are not guaranteed to
  // be thread safe across all engines ...
  //
  // We also leak them due to apparent teardown segfaults observed since cuDNN
  // version 9.10+
  ```

So: per-thread LRU, default cap ~2 GiB each, and the cache objects are deliberately leaked on (and after) cuDNN 9.10. Below the per-thread cap the cache simply grows with the number of distinct shapes that thread has seen (~200 KiB/plan), forever.

### Experiment 1 — it's cuDNN path, scales with #distinct shapes

To separate "shape-keyed cache" from "allocator retains bigger buffers", we hold #calls (1500) and the size band [256,512] fixed, draw lengths in random order, and vary only the number of **distinct** lengths. Per-shape signal = `(distinct=many) − (distinct=few)`.

| config (`Conv1d` ×3 kernels 3/7/15) | distinct | host RSS Δ | `many−few` |
|---|---:|---:|---:|
| cuda, cudnn=True, benchmark=False | 200 | +26.5 MiB | **+16.9** |
| cuda, cudnn=True, benchmark=False |   4 |  +9.6 MiB | |
| cuda, **cudnn=False** (native conv) | 200 | +16.8 MiB | **+0.6** |
| cuda, **cudnn=False** (native conv) |   4 | +16.2 MiB | |
| cuda, cudnn=True, **benchmark=True** | 200 | +18.5 MiB | **+17.6** |
| cuda, cudnn=True, inputs **bucketed to mult. of 64** | 5 | +0.5 MiB | — |

→ growth scales with #distinct shapes; `cudnn.enabled=False` kills it (≈+0.6); bucketing shapes kills it (+0.5).

### Experiment 2 — two aggravating factors

Same workload with 200 distinct shapes, but we vary threads / LRU env var:

| config | host RSS Δ |
|---|---:|
| 1 thread, default limit | +26.6 MiB |
| **8 threads**, default limit | **+102.8 MiB** (~4×) |
| 8 threads, `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT=25` | +23.1 MiB |
| 1 thread, `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT=25` | +10.3 MiB |

→ the cache is per-thread (more threads ⇒ proportionally more host memory for *same* workload), and env cap bounds it. So server with 40-thread inference pool and few thousand distinct shapes will take many GiB with default limit — this matches slow OOM we saw on production.

### Mitigations that work today (no PyTorch change)

1. `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT=<small>` (e.g. a few hundred) — bounds the per-thread plan cache.
2. Cap the inference thread-pool size (fewer thread-local caches).
3. Pad/bucket conv input lengths to a small fixed set of shapes (what we shipped) — fewer distinct plans regardless of threads.

### Possible fixes upstream (just ideas — I didn't built/tested them)

The per-thread LRU bounds *one* thread; nothing bounds the aggregate. Options:
- a process-wide budget (cap total cached plans across all thread-local caches, e.g. an atomic counter + global LRU eviction), instead of / in addition to the per-thread limit;
- a smaller / clearly documented default (the "2 GiB" estimate is *per thread* and easy to miss);
- a way to reclaim caches of terminated threads despite the 9.10+ teardown segfault — e.g. register each thread-local cache in a global registry and free them at a safe point (interpreter shutdown / explicit `torch.cuda` API) rather than relying on thread-exit destructors;
- an explicit `torch.backends.cudnn` API to clear the plan cache (handy for long-lived servers between workloads).

I am happy to test patches using repro below.

### Repro

Shape / attribution (only needs `torch`; self-spawns one process per config, prints the Experiment-1 table):

```python
"""Host RSS grows with #distinct conv shapes (pytorch#104701). Run: python repro.py (Linux)."""
import json, os, random, subprocess, sys

def host_rss_mib():
    with open("/proc/self/status") as f:
        for line in f:
            if line.startswith("VmRSS:"):
                return int(line.split()[1]) / 1024
    return -1.0

def child(cfg):
    import torch, torch.nn as nn
    torch.backends.cudnn.enabled = bool(cfg["cudnn"]); torch.backends.cudnn.benchmark = bool(cfg["benchmark"])
    dev = torch.device(cfg["device"])
    convs = [nn.Conv1d(64, 64, k, padding=k // 2).to(dev).eval() for k in (3, 7, 15)]
    rng = random.Random(0)
    pool = rng.sample(range(cfg["lo"], cfg["hi"] + 1), cfg["distinct"])
    if cfg["bucket"]:
        pool = [((t + cfg["bucket"] - 1) // cfg["bucket"]) * cfg["bucket"] for t in pool]
    def run(t):
        x = torch.randn(1, 64, t, device=dev)
        with torch.no_grad():
            for c in convs: c(x)
        if dev.type == "cuda": torch.cuda.synchronize()
    for w in (8, 12): run(w)
    if dev.type == "cuda": torch.cuda.synchronize()
    r0 = host_rss_mib()
    for _ in range(cfg["calls"]): run(rng.choice(pool))
    print(f"RSSDELTA {host_rss_mib() - r0:.1f} {len(set(pool))}")

def main():
    if len(sys.argv) > 1 and sys.argv[1] == "--child":
        return child(json.loads(sys.argv[2]))
    import torch
    base = {"lo": 256, "hi": 512, "calls": 1500, "bucket": 0}
    cfgs = []
    if torch.cuda.is_available():
        cfgs += [
            ("cuda cudnn=1           distinct=200", {**base,"device":"cuda","cudnn":1,"benchmark":0,"distinct":200}),
            ("cuda cudnn=1           distinct=4  ", {**base,"device":"cuda","cudnn":1,"benchmark":0,"distinct":4}),
            ("cuda cudnn=1 bucket=64 distinct=200", {**base,"device":"cuda","cudnn":1,"benchmark":0,"distinct":200,"bucket":64}),
            ("cuda cudnn=0 (native)  distinct=200", {**base,"device":"cuda","cudnn":0,"benchmark":0,"distinct":200}),
            ("cuda cudnn=0 (native)  distinct=4  ", {**base,"device":"cuda","cudnn":0,"benchmark":0,"distinct":4}),
            ("cuda cudnn=1 bench=1   distinct=200", {**base,"device":"cuda","cudnn":1,"benchmark":1,"distinct":200}),
        ]
    cfgs += [("cpu (oneDNN)           distinct=200", {**base,"device":"cpu","cudnn":1,"benchmark":0,"distinct":200})]
    print(f"torch={torch.__version__} cuda={torch.version.cuda} cudnn={torch.backends.cudnn.version()}")
    for label, cfg in cfgs:
        out = subprocess.run([sys.executable, __file__, "--child", json.dumps(cfg)], capture_output=True, text=True)
        line = next((l for l in out.stdout.splitlines() if l.startswith("RSSDELTA")), None)
        print(f"{label:36s}  {line.split()[2]:>4s} distinct  {float(line.split()[1]):+7.1f} MiB" if line else f"{label}: ERR {out.stderr[-120:]}")

if __name__ == "__main__":
    os.environ.setdefault("CUDA_MODULE_LOADING", "LAZY")
    main()
```

**Thread multiplication + env cap** (Experiment 2): run the same workload from N threads, optionally with `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT` set:

```python
import os, random, threading, torch, torch.nn as nn
def rss():  # MiB
    return int(next(l for l in open("/proc/self/status") if l.startswith("VmRSS:")).split()[1]) / 1024
dev = torch.device("cuda"); torch.backends.cudnn.benchmark = False
convs = [nn.Conv1d(64,64,k,padding=k//2).to(dev).eval() for k in (3,7,15)]
pool = random.Random(0).sample(range(256,513), 200)
def run(t):
    with torch.no_grad():
        for c in convs: c(torch.randn(1,64,t,device=dev))
    torch.cuda.synchronize()
run(8); torch.cuda.synchronize(); r0 = rss()
N, CALLS = 8, 1600                                   # set N=1 to compare; set the env var to cap
def w(s):
    rng = random.Random(s)
    for _ in range(CALLS // N): run(rng.choice(pool))
ts = [threading.Thread(target=w, args=(i,)) for i in range(N)]
[t.start() for t in ts]; [t.join() for t in ts]; torch.cuda.synchronize()
print(f"threads={N} limit={os.environ.get('TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT','default')} dRSS={rss()-r0:+.1f}MiB")
```

Believe it helps. Best regards
Draqun

**Comment by eqy** [comment]
What's the issue with using a thread pool of workers? Memory usage is already effectively unbounded if the total number of threads is unbounded due to leaked cuBLAS/cuDNN handles etc

**Comment by Draqun** [comment]
Hi. @eqy thanks for reply.
Fair point on the thread pool — to be clear, we're not arguing for an unbounded number of threads. You're right that unbounded threads already blow up memory via per-thread cuBLAS/cuDNN handles (and this plan cache is itself leaked per-thread since 9.10), and that a bounded pool bounds the per-thread multiplier. I'll drop that framing — it wasn't the core of the problem.

The axis that was actually unbounded for us is the number of distinct input shapes, independent of threads. It reproduces single-threaded:

- 1500 calls, lengths drawn from `[256, 512]`, `benchmark=False`:
- 200 distinct shapes → +~17 MiB host RSS
- 4 distinct shapes → ~flat
- It grows roughly linearly with the number of distinct shapes and only stops at the default per-thread LRU cap (`TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT` = 10000 entries ≈ 2 GiB).

Our production server had a bounded thread pool (a few tens of threads), but the input `seq_len` was effectively unbounded (variable-length sequences). With `benchmark=False` and far fewer than 10000 distinct shapes, the LRU never evicted, so each thread's cache grew linearly with distinct shapes (x the bounded thread count) until the node OOM'd (there was no container memory limit).

So the surprising part for us wasn't "per-thread × threads" — it's that with variable input shapes the per-thread cache grows toward a ~2 GiB/thread default before any eviction kicks in, and there's no process-wide bound across the (bounded) pool. Concretely:

1. Is the 10000-entry (~2 GiB/thread) default intended to be that high? For variable-shape inference a lot of host memory accrues before the LRU does anything.
2. Would a short doc note help — e.g. "variable convolution input shapes cause host RSS to grow up to the per-thread plan-cache cap; mitigate by padding/bucketing shapes or lowering `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT`"? That alone would have shortcut our investigation.

We resolved it by padding/bucketing the conv input length to a small fixed set of shapes (≈ +0.5 MiB for 200 distinct lengths collapsing to 5 buckets), which sidesteps the cache growth regardless of thread count. Posting the distinct-shape framing in case it helps others doing variable-length inference.

Best regards.
Draqun
